### DISCUSSION TITLE
EVP_PKEY_generate()

### DISCUSSION THREAD

**Discussion by shaikhus12122864** [openssl/openssl#24011] [Q&A]
Is there any function in openssl to calculate CRT parameters for RSA private key

**Comment by kulkarniamit** [comment]
You can use the mappings specified in https://github.com/openssl/openssl/discussions/23506#discussioncomment-8393705 to get the CRT parameters of a given RSA private key.
Sample:
```c
#include <openssl/evp.h>
#include <openssl/core_names.h>
#include <openssl/rsa.h>

#define RSA_KEY_SIZE 2048


void print_parameter(const EVP_PKEY *pkey, const char *key_name) {
    BIGNUM *param = NULL;
    char *str = NULL;
    if (EVP_PKEY_get_bn_param(pkey, key_name, &param)) {
        str = BN_bn2dec(param);
        fprintf(stdout, "%s: %s\n", key_name, str);
        OPENSSL_free(str);
        BN_free(param);
    }
    else {
        fprintf(stderr, "Failed to fetch %s\n", key_name);
    }
}

int main() {
    EVP_PKEY *pkey = NULL;
    pkey = EVP_RSA_gen(RSA_KEY_SIZE);
    if (pkey == NULL) {
        fprintf(stderr, "Failed to generate RSA key\n");
        return 1;
    }
    print_parameter(pkey, OSSL_PKEY_PARAM_RSA_N);
    print_parameter(pkey, OSSL_PKEY_PARAM_RSA_FACTOR1);
    print_parameter(pkey, OSSL_PKEY_PARAM_RSA_FACTOR2);
    print_parameter(pkey, OSSL_PKEY_PARAM_RSA_D);
    print_parameter(pkey, OSSL_PKEY_PARAM_RSA_EXPONENT1);
    print_parameter(pkey, OSSL_PKEY_PARAM_RSA_EXPONENT2);
    print_parameter(pkey, OSSL_PKEY_PARAM_RSA_COEFFICIENT1);

    EVP_PKEY_free(pkey);
    return 0;
}
```

**Reply by shaikhus12122864** [reply]
Thanks, I have executed this code. Output seems to produce all CRT Parameters in the key.

Just one ask. Does  EVP_RSA_gen method populate all parameters by default without any specification from developer.

**Reply by t8m** [reply]
Generation of a new key always computes all the parameters.

**Comment by t8m** [comment]
> Is there any function in openssl to calculate CRT parameters for RSA private key

Do you need calculating CRT parameters for an existing key data which has only p, q and d? Or is this just asking about how to retrieve CRT parameters from an RSA EVP_PKEY generated by OpenSSL?

**Reply by shaikhus12122864** [reply]
Yeah, you got me right. I need to know are  there any methods available for calculating CRT Parameters from p, q& d

**Reply by t8m** [reply]
The OpenSSL 3.3 includes this change: https://github.com/openssl/openssl/blob/openssl-3.3/CHANGES.md?plain=1#L69

So with this version it is possible to import a key with just p, q and d params with EVP_PKEY_fromdata() and if you also add OSSL_PKEY_PARAM_RSA_DERIVE_FROM_PQ, the CRT params will be derived.
